pkg/waiter/manager.go

```diff
@@ -0,0 +1,255 @@
+package waiter
+
+import (
+	"context"
+	"errors"
+
+	"github.com/stackrox/rox/pkg/sync"
+)
+
+const (
+	// The default number of attempts to re-generate an ID on collision.
+	defaultMaxCollisions = 5
+)
+
+var (
+	// ErrTooManyCollisions the ID generator produced too many non-unique IDs.
+	ErrTooManyCollisions = errors.New("too many id collisions")
+
+	// ErrManagerShutdown manager has been shutdown and therefore cannot process requests.
+	ErrManagerShutdown = errors.New("manager is shutdown")
+
+	errKeyExists = errors.New("id exists")
+	errNoExists  = errors.New("id does not exist")
+)
+
+type response[T any] struct {
+	id   string
+	data T
+	err  error
+}
+
+type options struct {
+	idGenerator   IDGenerator
+	maxCollisions int
+}
+
+// Option defines a functional option for configuring a manager.
+type Option func(option *options)
+
+func noopOptionFunc(_ *options) {}
+
+// WithIDGenerator is a functional option used to change a managers default ID generator.
+func WithIDGenerator(gen IDGenerator) Option {
+	if gen == nil {
+		return noopOptionFunc
+	}
+
+	return func(options *options) {
+		options.idGenerator = gen
+	}
+}
+
+// WithMaxCollisions is a functional option used to change the the max number of collisions
+// that can occur before waiter creation fails.
+func WithMaxCollisions(num int) Option {
+	if num < 1 {
+		return noopOptionFunc
+	}
+
+	return func(options *options) {
+		options.maxCollisions = num
+	}
+}
+
+// Manager builds waiters and delivers messages to waiters from async publishers.
+//
+//go:generate mockgen-wrapper
+type Manager[T any] interface {
+	// Start spawns a goroutine that will run forever or until ctx.Done() delivering
+	// messages to waiters.
+	Start(ctx context.Context)
+
+	// Send sends data and err to the waiter with the provided id.
+	Send(id string, data T, err error) error
+
+	// NewWaiter creates a waiter with a unique ID.
+	NewWaiter() (Waiter[T], error)
+}
+
+type managerImpl[T any] struct {
+	// idGenerator is responsible for generating unique waiter IDs.
+	idGenerator IDGenerator
+
+	// waiters holds a waiter's id and the channel to send responses on.
+	waiters map[string]chan *response[T]
+
+	// waitersMu ensures only one goroutine is manipulating the waiters map at a time.
+	waitersMu sync.Mutex
```

Comment:
nit: could use a `sync.RWMutex` instead to make the intent more clear which functions make changes / which read.

Reply:
I considered that! Besides the `len()` function (only for tests), all the other functions manipulate the waiters in some way so changed it to `sync.Mutex`